### DISCUSSION TITLE
IStorageProvider.OpenFilePickerAsync  randomly not showing the dialog window.

### DISCUSSION THREAD

**Discussion by dfkeenan** [AvaloniaUI/Avalonia#13484] [Q&A]
Hi,

I have been upgrading an Avalonia application to v11. I am trying to use the new `IStorageProvider` API. I am using `ReactiveUI` and calling the `OpenFilePickerAsync ` from within a `ReactiveCommand`. Seemingly randomly the `Task` returned from the `OpenFilePickerAsync` method awaits forever and no dialog window is shown. 

The project I am working on has a lot of things going on. `ReactiveUI`, `Autofac` dependency injection to name a few. So far I have been unable to reproduce the issue in a smaller/simpler project.

I am working on Windows 11 desktop.

Any help on producing a repro (or fixing the issue) would be greatly appreciated.

Thanks,
dfkeenan

**Comment by ibrahim324** [comment]
I am working with StorageProvider as well. My issue was that the Dialog didn't show up at all. I resorted to writing an async Click method that sets the respective ViewModel property. My view is a ReactiveWindow<MainWindowViewModel>.
I am fine with this as it is, because the view is now responsible for opening another view (the dialog), though I am open for suggestions.
```
private async void ButtonClick(object sender, RoutedEventArgs e)
{
    var provider = this.StorageProvider;
    var options = new FolderPickerOpenOptions{Title="Pick a Directory", AllowMultiple=false};
    var dialog = await provider.OpenFolderPickerAsync(options);
    this.ViewModel.Directories = dialog;
}
```

further you might want to take a look [here](https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/DialogsPage.xaml.cs). Good luck!

**Reply by thevortexcloud** [reply]
If you are using ReactiveUI there are far better ways to do this. EG you can use a standard command to launch an interaction which returns the dialogue result. That way it makes a full round trip, from your view (the command), to your view model (which fires the interaction if it wants to, which notifies the view), back to your view, and then finally back to your view model with the final result. The other big advantage of this is your view model can prompt for a dialogue result on its own without being initiated by the view.

It's a lot more work to set this all up, but it lets you do things like this using standard methods. You can define a bunch of generic interactions (EG open file dialogue) and put them in your base view model and put some generic interaction handlers into your base view. That way in most cases you only need to write this once and it will be available in all your views and view models. 

The music store tutorial app for Avalonia explains how to use interactions as well as has some examples. But if it helps, think of interactions as blocking events that can pass data in both directions. Due to the whole subscription system, you can also decouple completely from your views (you just register some other handler for the interaction) and do full unit tests of your view models.

https://docs.avaloniaui.net/docs/tutorials/music-store-app/opening-a-dialog

**Reply by ibrahim324** [reply]
interesting, I'll take a look!

**Reply by ibrahim324** [reply]
@thevortexcloud I have not worked through the tutorial, but worked through the [documentation for Interactions](https://www.reactiveui.net/docs/handbook/interactions/index.html), and it works brillantly!

My ViewModel: 
```
public class MainWindowViewModel : ReactiveObject
{
    private readonly Interaction<Unit, string> selectDir;
    public Interaction<Unit, string> SelectDirectoryInteraction => selectDir;

    ObservableAsPropertyHelper<string> selectedDirectory;
    public string SelectedDirectory => selectedDirectory.Value; 

    public ReactiveCommand<Unit, string> GetDirectoryCommand{get; private set;}
    public MainWindowViewModel()
    {
        GetDirectoryCommand = ReactiveCommand.CreateFromTask(getDirectory);
        
        this.selectDir = new Interaction<Unit, string>();
        
        this.selectedDirectory = 
            GetDirectoryCommand
            .ToProperty(this, f => f.SelectedDirectory);
    }

    private async Task<string> getDirectory()
    {
        var dir = await selectDir.Handle(Unit.Default);
        if(string.IsNullOrWhiteSpace(dir)) 
            return string.Empty;
        return dir;
    }
}
```

My View: 
```
public MainWindow()
    {
        ViewModel = new MainWindowViewModel();
        InitializeComponent();
        this.WhenActivated(d => 
        {
            ViewModel.SelectDirectoryInteraction.RegisterHandler(async interaction => 
            {
                var storage = this.StorageProvider;
                var options = new FolderPickerOpenOptions{Title="Choose a Directory!", AllowMultiple=false};
                var dirs = await storage.OpenFolderPickerAsync(options);
                var dir = dirs.FirstOrDefault()?.Path.AbsolutePath;
                interaction.SetOutput(dir);
            });

            this.BindCommand(ViewModel, vm => vm.GetDirectoryCommand, v => v.Button)
                .DisposeWith(d);

            this.OneWayBind(ViewModel, vm => vm.SelectedDirectory, v => v.TextBlock.Text)
                .DisposeWith(d);
        });            
    }
```

**Reply by thevortexcloud** [reply]
Glad to hear. That said you can simplify this somewhat. You don't need to explicitly declare the backing variable for the interaction in the view model. You can just do a `get` only property with a default value, EG ` public Interaction<Unit, string> SelectDirectoryInteraction {get;} = new Interaction<Unit,string>();`

You can also make your view inherit from one of the  `Reactive` view base types. EG `ReactiveWindow<MainWindowViewModel>`. This gives you a strongly typed `ViewModel` property among other things. I am actually somewhat surprised your view is working as it's not inheriting from any of the Avalonia types, or any of the ReactiveUI types which inherit from an Avalonia type, nor is it a partial class.

Also is there a particular reason you are not binding in XAML? It saves you from having to do things like `OneWayBind` and keeps your code behind far cleaner.

**Reply by ibrahim324** [reply]
Thanks for the suggestions! My view does inherit from ReactiveWindow<MainWindowViewModel>, however I have not declared it in xaml. Still seems to work though.
Regarding the Bindings : I just follow RxUI's [documentation](https://www.reactiveui.net/docs/handbook/data-binding/); apparently xaml-based Bindings can leak memory, though I agree that binding in xaml keeps the view clean

**Reply by thevortexcloud** [reply]
>  My view does inherit from ReactiveWindow, 

Ah my bad. I misread your constructor as your class declaration somehow.

>Regarding the Bindings : I just follow RxUI's [documentation](https://www.reactiveui.net/docs/handbook/data-binding/); apparently xaml-based Bindings can leak memory, though I agree that binding in xaml keeps the view clean

Fair enough. Although based on what I am reading that only applies if you are binding to things that don't implement `INotifyPropertyChanged`

**Reply by jedidja** [reply]
+1 to @thevortexcloud's response - stumbled across this thread while trying to open a dialog from the VM :)

**Reply by thevortexcloud** [reply]
I will also add my response in this thread to someone who asked a similar question, and the thread also contains information about how to do it without ReactiveUI

https://github.com/AvaloniaUI/Avalonia/discussions/12475#discussioncomment-6663377

**Comment by thevortexcloud** [comment]
Is there a particular reason you are trying to directly open a file dialogue from your view model? It's generally not considered good practice to couple your view model directly to presentation concerns (i.e assuming the user will select a file via a dialogue instead of some other method). As the view model does not really care about how it receives the data, just that it does. To quote Microsoft on the MVVM pattern:

>In addition to understanding the responsibilities of each component, it's also important to understand how they interact. At a high level, the view "knows about" the view model, and the view model "knows about" the model, but the model is unaware of the view model, and the view model is unaware of the view. Therefore, the view model isolates the view from the model, and allows the model to evolve independently of the view.

https://learn.microsoft.com/en-us/dotnet/architecture/maui/mvvm

This also directly couples your viewmodel to Avalonia, which may not be desirable if you ever need to port your code to another MVVM framework. [As mentioned in my other comment](https://github.com/AvaloniaUI/Avalonia/discussions/13484#discussioncomment-7478892), ReactiveUI has a built in solution for this.

**Comment by dfkeenan** [comment]
How I am getting my hands on an `IStorageProvider` may be questionable and I might change it. 

But doesn't really explain why when the command executes I have an `IStorageProvider`, `IStorageProvider.CanOpen ` is `true` and the call to `OpenFilePickerAsync` returns a `Task<IReadOnlyList<IStorageFile>>`. Sometime I get a window open and sometimes it just awaits "forever". Seems to work the first time but fails on a some random subsequent button click. Could be the 2nd or the 5th. Quite confusing.

**Reply by thevortexcloud** [reply]
It's _probably_ a threading issue. Most UI frameworks are fairly picky about who can present things onto the screen.  Have you tried always forcing it onto the UI thread using `Dispatcher`? Tasks generally provide no guarantee of what thread code is run on, and may return on an entirely different thread than the one that started the task (which is why using things like mutexes with tasks is a nightmare).

It might be something else entirely though.

That said I have personally never run into this problem when opening file pickers via my views or bouncing the request from my view model to the view using interactions.

**Reply by maxkatz6** [reply]
Windows storage provider opens new dialog in a new thread. There shouldn't be any blocking issues: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32StorageProvider.cs#L76

**Comment by dfkeenan** [comment]
I think I fixed it.

I wasn't doing my startup code in the `Application.OnFrameworkInitializationCompleted` method, so I tried moving it there. Then I started getting invalid thread exception when creating the window.  I was using the `System.CommandLine` package to get some arguments. Creating the window inside the command handler seems to cause it. I removed that logic for now. And the problem seems to have gone away.
